File: Java/Euler (Java)/src/problems/P001xP050/P010.java
package problems.P001xP050;

import java.util.List;

import util.Primes;
import base.Problem;

/**
 * P010<br>
 * The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.<br>
 * Find the sum of all the primes below two million.<br>
 * Ans: 142913828922
 */
public class P010 extends Problem
{
   public P010()
   {
      variations = 2;
   }

   @Override
   protected String execute(int variation)
   {
      switch (variation)
      {
         case 0:
            return bruteForce() + "";
         case 1:
         case -1:
            return optimized() + "";
      }
      return null;
   }

   /**
    * Just check every number!
    * 
    * @return
    */
   private long bruteForce()
   {
      long sum = 0;
      for (int ii = 2; ii < 2000000; ii++)
      {
         if (Primes.isPrime(ii))
            sum += ii;
      }
      return sum;
   }

   /**
    * Use the sieve of erosthenes and just add!
    * 
    * @return
    */
   private long optimized()
   {
      long sum = 0;
      List<Integer> primes = Primes.EratosthenesPrimesUpTo(2000000);
      for (int p : primes)
         sum += p;
      return sum;
   }
}

File: C++/Euler (C++)/src/problems/P001xP010/P010.cpp
#include "../../base/problem.h"
#include "../../util/prime.h"

/**
 * P010<br>
 * The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.<br>
 * Find the sum of all the primes below two million.<br>
 * Ans: 142913828922
 */
class P010: public Problem
{
   public:
      P010()
      {
         problem_number = 10;
         variations = 2;
      }
      ~P010()
      {
      }

      std::string execute(int variation)
      {
         switch (variation)
         {
            case 0:
               return to_string(bruteForce());
            case 1:
            case -1:
               return to_string(optimized());
            default:
               return std::string("unavailable");
         }
         return NULL;
      }

   private:

      /**
       * Just check every number!
       * @return
       */
      long bruteForce()
      {
         long sum = 0;
         for (int ii = 2; ii < 2000000; ii++)
         {
            if (prime::isPrime(ii))
               sum += ii;
         }
         return sum;
      }

      /**
       * Use the sieve of erosthenes and just add!
       * @return
       */
      long optimized()
      {
         long sum = 0;
         std::vector<int>* primes = prime::EratosthenesPrimesUpTo(2000000);
         for (std::vector<int>::iterator p = primes->begin(); p != primes->end(); ++p)
            sum += *p;
         delete primes;
         return sum;
      }
};
